### PR TITLE
persistent volumes in group - handle pvc for create, update and delete of pod

### LINKED ISSUE #29: Controller incorrectly captures pod deletion time
**Is this a BUG REPORT or FEATURE REQUEST?**:
 /kind bug

**What happened**:
Controller incorrectly captures pod deletion time [here](https://github.com/vmware/purser/blob/0fadf138fc11d05604002ef59edbf9792a305bc7/pkg/controller/controller/group_controller.go#L260)

**What you expected to happen**:
It should capture deletion time from `DeletionTimeStamp` of Object metadata, but the object might have been completely removed from etcd. So deletion time can be set approximately as the time when the controller captured this event.
Assumption: Delay between the actual deletion occurring and controller capturing event is not significant compared to pod life time.

### PR DESCRIPTION
**Pull request by kreddyj**
**What this PR does / why we need it**:
1. fix: controller - deletion time of pod is set incorrectly [here](https://github.com/vmware/purser/blob/0fadf138fc11d05604002ef59edbf9792a305bc7/pkg/controller/controller/group_controller.go#L260)

2. Also added few todos to handle cases of missing events.

3. Saving capture time (may be needed in future to handle update case)

4. Persistent Volumes in group - handle pvc for create, update and delete of pod.

Fixes #29 and partly solves issue #30

### CODE REVIEW

**Comment by kreddyj** [comment]
@gurusreekanthc updated the pod deletion time.

File: pkg/controller/controller/group_controller.go [hunk 1]
```diff
@@ -301,3 +322,53 @@ func isPodBelongsToGroup(group *crd.Group, pod *api_v1.Pod) bool {
...
+
+		return &crd.PersistentVolumeClaim{
+			Name:                pvc.GetObjectMeta().GetName(),
+			VolumeName:          pvc.Spec.VolumeName,
+			RequestSizeInGB:     []float64{utils.BytesToGB(request.Value())},
+			CapacityAllotedInGB: []float64{utils.BytesToGB(capacity.Value())},
+			BoundTimes:          []meta_v1.Time{},
+			UnboundTimes:        []meta_v1.Time{},
+		}
+	}
```
Comment (hemani19) [inline pkg/controller/controller/group_controller.go]: Use of too many `if-else` constructs is generally not a good practise. 

I would recommend breaking this into something like: 

``` go
if err != nil {
		if errors.IsNotFound(err) {
			log.Printf("Persistent Volume Claim %s not found\n", claimName)
		}

		if statusError, isStatus := err.(*errors.StatusError); isStatus {
			log.Printf("Error getting persistence volume Claim %s : %v\n", claimName, statusError.ErrStatus.Message)
		}
		return nil
	}

	request := pvc.Spec.Resources.Requests["storage"].DeepCopy()
	capacity := pvc.Status.Capacity["storage"].DeepCopy()

	return &crd.PersistentVolumeClaim{
		Name:                pvc.GetObjectMeta().GetName(),
		VolumeName:          pvc.Spec.VolumeName,
		RequestSizeInGB:     []float64{utils.BytesToGB(request.Value())},
		CapacityAllotedInGB: []float64{utils.BytesToGB(capacity.Value())},
		BoundTimes:          []meta_v1.Time{},
		UnboundTimes:        []meta_v1.Time{},
	}
``` 

Probably, even abstracting the error handling into a separate method something like `checkAndLogErr()`.

File: pkg/controller/utils/unitConversions.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package utils
+
+// BytesToGB converts from bytes(int64) to GB(float64)
+func BytesToGB(val int64) float64 {
+	return float64(val) / (1024.0 * 1024.0 * 1024.0)
+}
```
Comment (hemani19) [inline pkg/controller/utils/unitConversions.go]: Good Abstraction 👍

**Comment by hemani19** [comment]
@kreddyj, good job. Just one small suggestion from my side otherwise I am good with it. 👍

File: pkg/controller/controller/group_controller.go (hunk 1 above)
Comment (kreddyj) [reply pkg/controller/controller/group_controller.go]: Done.

File: pkg/controller/utils/unitConversions.go (hunk 2 above)
Comment (kreddyj) [reply pkg/controller/utils/unitConversions.go]: Its sreekanth's abstraction in plugin side, just copied that into controller too. :smile:

File: pkg/controller/controller/group_controller.go [hunk 3]
```diff
@@ -301,3 +322,53 @@ func isPodBelongsToGroup(group *crd.Group, pod *api_v1.Pod) bool {
...
+			podVolumes[claimName] = collectPersistentVolumeClaim(claimName)
+			podVolumes[claimName].BoundTimes = append(podVolumes[claimName].BoundTimes, pod.GetCreationTimestamp())
+		}
+	}
+
+	return podVolumes
+}
+
+func collectPersistentVolumeClaim(claimName string) *crd.PersistentVolumeClaim {
+	pvc, err := ClientSetInstance.CoreV1().PersistentVolumeClaims("").Get(claimName, meta_v1.GetOptions{})
```
Comment (gurusreekanthc) [inline pkg/controller/controller/group_controller.go]: Pass same namespace as pod. Assumption: Pod can't have pvc from other namespace. Please verify the same.

File: pkg/controller/crd/group_crd.go [hunk 4]
```diff
@@ -60,10 +60,11 @@ type GroupSpec struct {
...
 type PodDetails struct {
-	Name       string
-	StartTime  meta_v1.Time
-	EndTime    meta_v1.Time
-	Containers []*Container
+	Name          string
+	StartTime     meta_v1.Time
+	EndTime       meta_v1.Time
+	Containers    []*Container
+	PodVolumesMap map[string]*PersistentVolumeClaim
```
Comment (gurusreekanthc) [inline pkg/controller/crd/group_crd.go]: PodVolumesMap >> rename to PodVolumes or something similar.
We don't need to specify the type in name.

File: pkg/controller/controller/payload.go [hunk 5]
```diff
@@ -31,4 +35,5 @@ type Payload struct {
 	ResourceType string `json:"resourceType"`
 	CloudType    string `json:"cloudType"`
 	Data         string `json:"data"`
+	CaptureTime  meta_v1.Time
```
Comment (gurusreekanthc) [inline pkg/controller/controller/payload.go]: What is the use of CaptureTime?

File: pkg/controller/controller/group_controller.go (hunk 3 above)
Comment (kreddyj) [reply pkg/controller/controller/group_controller.go]: Done

File: pkg/controller/crd/group_crd.go (hunk 4 above)
Comment (kreddyj) [reply pkg/controller/crd/group_crd.go]: Done

**Comment by kreddyj** [comment]
@gurusreekanthc @hemani19 made changes based on your comments.

File: pkg/controller/controller/controller.go [hunk 6]
```diff
@@ -82,6 +86,8 @@ func Start(conf *config.Config) {
 		kubeClient = utils.GetClient()
 	}
 
+	ClientSetInstance = kubeClient
```
[the remark this answers is not in the document]
Comment (kreddyj) [reply pkg/controller/controller/controller.go]: ok

File: pkg/controller/controller/group_controller.go [hunk 7]
```diff
@@ -301,3 +322,66 @@ func isPodBelongsToGroup(group *crd.Group, pod *api_v1.Pod) bool {
...
+		CapacityAllotedInGB: []float64{utils.BytesToGB(capacity.Value())},
+		BoundTimes:          []meta_v1.Time{},
+		UnboundTimes:        []meta_v1.Time{},
+	}
+}
+
+func updatePodVolumes(podDetails *crd.PodDetails) {
+	pvMap := podDetails.PodVolumes
+	for claimName := range pvMap {
+		// If length of bound times is 1 more than unbound times it means that the pvc is still bound to pod.
```
[the remark this answers is not in the document]
Comment (kreddyj) [reply pkg/controller/controller/group_controller.go]: ok will do that

[the remark this answers is not in the document]
Comment (kreddyj) [reply pkg/controller/controller/group_controller.go]: @gurusreekanthc i made changes. function `updatePodVolumes` in `pkg/controller/controller/persistentVolume.go` takes care of when to bound and when to unbound a pvc.

File: pkg/controller/controller/controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kreddyj) [reply pkg/controller/controller/controller.go]: Done

File: pkg/controller/controller/persistentVolume.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+	// TODO: handle Case Resizing of PVC
+
+	podDetails.PodVolumes = podVolumes
+	return podDetails
+}
+
+func getActivePodVolumes(pod api_v1.Pod) map[string]*crd.PersistentVolumeClaim {
+	namespace := pod.GetNamespace()
+	podVolumes := map[string]*crd.PersistentVolumeClaim{}
```
Comment (gurusreekanthc) [inline pkg/controller/controller/persistentVolume.go]: podVolumes >> podVolumeClaims

**Comment by kreddyj** [comment]
@gurusreekanthc renamed the variable and did an end to end test. I added logs for capturing pvc events. In the test, expected logs(`Bounded new pvc: pvcName`) are generated.

**Pull request merged**